program/src/processor.rs

```diff
@@ -818,35 +819,93 @@ fn process_switch_vote(
     let accounts_iter = &mut accounts.iter();
 
     let proposal_info = next_account_info(accounts_iter)?;
+    let proposal_transaction_info = next_account_info(accounts_iter)?;
 
     check_proposal_exists(program_id, proposal_info)?;
 
     let mut proposal_data = proposal_info.try_borrow_mut_data()?;
     let proposal_state = bytemuck::try_from_bytes_mut::<Proposal>(&mut proposal_data)
         .map_err(|_| ProgramError::InvalidAccountData)?;
 
-    let clock = <Clock as Sysvar>::get()?;
-
-    // TODO: These checks do the same thing basically.
-    // Will be cleaned up when this processor is rearchitected.
-    if !proposal_state.cooldown_has_ended(&clock) {
-        return Err(PaladinGovernanceError::ProposalNotAccepted.into());
-    }
+    // Ensure the proposal was accepted.
     if proposal_state.status != ProposalStatus::Accepted {
         return Err(PaladinGovernanceError::ProposalNotAccepted.into());
     }
 
-    // Process the proposal instruction.
-    // TODO!
+    // Ensure the provided proposal transaction account has the correct address
+    // derived from the proposal.
+    if !proposal_transaction_info
+        .key
+        .eq(&get_proposal_transaction_address(
+            proposal_info.key,
+            program_id,
+        ))
+    {
+        return Err(PaladinGovernanceError::IncorrectProposalTransactionAddress.into());
+    }
 
-    // Set the proposal's status to processed.
-    proposal_state.status = ProposalStatus::Processed;
+    check_proposal_transaction_exists(program_id, proposal_transaction_info)?;
+
+    let mut proposal_transaction_state =
+        ProposalTransaction::try_from_slice(&proposal_transaction_info.try_borrow_data()?)?;
+
+    // Ensure the index is valid.
+    let instruction_index = instruction_index as usize;
+    if instruction_index >= proposal_transaction_state.instructions.len() {
+        return Err(PaladinGovernanceError::InvalidTransactionIndex.into());
+    }
+
+    let instruction = &proposal_transaction_state.instructions[instruction_index];
+
+    // Ensure the instruction has not already been executed.
+    if instruction.executed {
+        return Err(PaladinGovernanceError::InstructionAlreadyExecuted.into());
+    }
+
+    // Ensure the previous instruction has been executed.
+    if instruction_index > 0
+        && !proposal_transaction_state.instructions[instruction_index.saturating_sub(1)].executed
+    {
+        return Err(PaladinGovernanceError::PreviousInstructionHasNotBeenExecuted.into());
+    }
+
+    // Execute the instruction.
+    {
+        let (_governance_address, signer_bump_seed) = get_governance_address_and_bump_seed(
+            &proposal_state.governance_config.stake_config_address,
+            program_id,
+        );
+        let bump_seed = [signer_bump_seed];
+        let governance_signer_seeds = collect_governance_signer_seeds(
+            &proposal_state.governance_config.stake_config_address,
+            &bump_seed,
+        );
```

Comment:
I think this should be the treasury PDA, not the governance PDA. We want the treasury to own everything and be the piggybank for the reward distribution system. If we use the treasury, then it's easy to make a proposal with a system transfer, and `process_instruction` will sign with the piggybank directly. Does that make sense?

Reply:
I think so. So the treasury is the same PDA as I have it now, but it's owned by System?

Reply:
And as a result, I can fold the entire #18 into just a plain-old system transfer instruction like I'm doing with alice and bob here.

Reply:
Also, should there still be a treasury per governance config? Or just one global treasury?

Reply:
Yeah that's exactly what I was thinking. And it should be one treasury per governance config, so that it isn't possible to create another governance config on the program and steal from the global treasury

Reply:
Alrighty, give that a look then!